controllers/openstackmachine_controller.go

```diff
@@ -280,11 +289,132 @@ func (r *OpenStackMachineReconciler) reconcileDelete(scope scope.Scope, cluster
 	return ctrl.Result{}, nil
 }
 
+// reconcileFloatingAddressFromPool creates IPAddressClaims for the OpenStackMachine if the OpenStackMachine has a
+// FloatingAddressFromPool specified. It then associates the IPAddressClaim with the instance.
+func (r *OpenStackMachineReconciler) reconcileFloatingAddressFromPool(ctx context.Context, scope scope.Scope, openStackMachine *infrav1.OpenStackMachine, openStackCluster *infrav1.OpenStackCluster, instanceStatus *compute.InstanceStatus, instanceNS *compute.InstanceNetworkStatus) error {
+	if openStackMachine.Spec.FloatingAddressFromPool == nil {
+		return nil
+	}
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		return err
+	}
+	computeService, err := compute.NewService(scope)
+	if err != nil {
+		return err
+	}
+
+	poolRef := openStackMachine.Spec.FloatingAddressFromPool
+
+	claimName := names.GetFloatingAddressIPAddressClaimName(openStackMachine.Name)
+	claim := &ipamv1.IPAddressClaim{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Namespace: openStackMachine.Namespace, Name: claimName}, claim); err != nil {
+		if apierrors.IsNotFound(err) {
+			claim = &ipamv1.IPAddressClaim{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      claimName,
+					Namespace: openStackMachine.Namespace,
+					Labels: map[string]string{
+						clusterv1.ClusterNameLabel: openStackCluster.Labels[clusterv1.ClusterNameLabel],
+					},
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: openStackMachine.APIVersion,
+							Kind:       openStackMachine.Kind,
+							Name:       openStackMachine.Name,
+							UID:        openStackMachine.UID,
+						},
+					},
+				},
+				Spec: ipamv1.IPAddressClaimSpec{
+					PoolRef: *poolRef,
+				},
+			}
+			if err := r.Client.Create(ctx, claim); err != nil {
+				return err
+			}
+			scope.Logger().Info("Created IPAddressClaim", "name", claim.Name)
+		} else {
+			return err
+		}
+	}
+
+	if claim.Status.AddressRef.Name != "" {
+		address := &ipamv1.IPAddress{}
+		addressKey := client.ObjectKey{Namespace: openStackMachine.Namespace, Name: claim.Status.AddressRef.Name}
+
+		if err := r.Client.Get(ctx, addressKey, address); err != nil {
+			return err
+		}
+
+		instanceAddresses := instanceNS.Addresses()
+		for _, instanceAddress := range instanceAddresses {
+			if instanceAddress.Address == address.Spec.Address {
+				openStackMachine.Status.FloatingAddressFromPoolReady = pointer.Bool(true)
+				conditions.MarkTrue(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition)
+				return nil
+			}
+		}
+
+		fip, err := networkingService.GetFloatingIP(address.Spec.Address)
+		if err != nil {
+			return err
+		}
+		if fip == nil {
+			conditions.MarkFalse(openStackMachine, infrav1.FloatingAddressFromPoolReadyCondition, infrav1.FloatingAddressFromPoolErrorReason, clusterv1.ConditionSeverityError, "floating IP does not exist")
+			return errors.New("floating IP does not exist in ")
```

Comment:
Looks like incomplete message? It'd probably be useful to mention the address itself.